### PR TITLE
Add a TimestampConverter that can be used to convert nanoTime to Timestamp.

### PR DESCRIPTION
**Pull request by bogdandrutu**

### CODE REVIEW

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    return new Time(timestamp, 0);
+  }
+
+  /**
+   * Returns the {@code Time} with the given {@code nanoTime}.
+   *
+   * @param nanoTime the nano time associated with this {@code Time}.
+   * @return the {@code Time} with the given {@code nanoTime}.
+   */
+  static Time withNanoTime(long nanoTime) {
```
Comment (sebright) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: How about something like `fromNanoTime`?  "with" makes it sound like there is some existing object or action that the method is adding something to.

> @param nanoTime the nano time associated with this {@code Time}.

I also think this phrase could be clearer, maybe "the nano time to be converted to a {@code Time}".

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+package com.google.instrumentation.trace;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.instrumentation.common.Timestamp;
+import com.google.instrumentation.common.TimestampFactory;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * This class provides a mechanism for converting {@link Time} values to {@link Timestamp}. An
+ * instance of this class in initialized with a pair of timestamp/nanoTime measurements that wre
```
Comment (sebright) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: s/wre/are

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+   *
+   * @param timestamp the timestamp associated with this {@code Time}.
+   * @return the {@code Time} with the given {@code timestamp}.
+   */
+  static Time withTimestamp(Timestamp timestamp) {
+    return new Time(timestamp, 0);
+  }
+
+  /**
+   * Returns the {@code Time} with the given {@code nanoTime}.
```
Comment (sebright) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: I think this comment should explain what the input represents, for example, "the given number of nanoseconds since ...".

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+   * converter.
+   *
+   * @param time value to convert.
+   * @return the {@code Timestamp} representation of the {@code time}.
+   */
+  Timestamp convert(Time time) {
+    if (time.hasTimestamp()) {
+      return time.getTimestamp();
+    }
+    return timestamp.addNanos(nanoTime - time.getNanoTime());
```
Comment (sebright) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: Is the subtraction reversed?

File: core/src/main/java/com/google/instrumentation/common/Timestamp.java [hunk 5]
```diff
@@ -13,18 +13,24 @@
...
  * <p>Use {@link TimestampFactory#now} to get the current timestamp since epoch
  * (1970-01-01T00:00:00Z).
  */
+@Immutable
 public final class Timestamp {
   private static final long MAX_SECONDS = 315576000000L;
   private static final int MAX_NANOS = 999999999;
   private static final long NUM_MILLIS_PER_SECOND = 1000L;
   private static final int NUM_NANOS_PER_MILLI = 1000 * 1000;
+  private static final long NUM_NANOS_PER_SECOND = NUM_MILLIS_PER_SECOND * NUM_NANOS_PER_MILLI;
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/common/Timestamp.java]: Nit: I think the calculation is easier to parse/check if it's reversed, i.e.

private static final long NUM_NANOS_PER_SECOND = NUM_NANOS_PER_MILLI * NUM_MILLIS_PER_SECOND;

File: core/src/main/java/com/google/instrumentation/common/Timestamp.java [hunk 6]
```diff
@@ -60,12 +64,14 @@ public static Timestamp create(long seconds, int nanos) {
 
   /**
    * Creates a new timestamp from given milliseconds.
+   *
+   * @return a new timestamp from
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/common/Timestamp.java]: Comment is truncated (from what?). Either delete the "from" or extend the sentence.

File: core/src/main/java/com/google/instrumentation/common/Timestamp.java [hunk 7]
```diff
@@ -90,6 +96,23 @@ public int getNanos() {
     return nanos;
   }
 
+  /**
+   * Returns the {@code Timestamp} calculated as this {@code Timestamp} plus {@code nanos}.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/common/Timestamp.java]: Suggest: Returns a {@code Timestamp} calculated as this {@code Timestamp} plus some number of nanoseconds.

File: core/src/main/java/com/google/instrumentation/common/Timestamp.java [hunk 8]
```diff
@@ -90,6 +96,23 @@ public int getNanos() {
...
+   *
+   * @param nanos the nanoseconds to be added to the current timestamp.
+   * @return the {@code Timestamp} calculated as this {@code Timestamp} plus {@code nanos}.
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/common/Timestamp.java]: s/the/a/

File: core/src/main/java/com/google/instrumentation/common/Timestamp.java [hunk 9]
```diff
@@ -90,6 +96,23 @@ public int getNanos() {
...
+   */
+  public Timestamp addNanos(long nanos) {
+    long newNanos = nanos + this.nanos;
```
Comment (a-veitch) [inline core/src/main/java/com/google/instrumentation/common/Timestamp.java]: Do you want to worry about overflow here? (the case where nanos = Long.MAX_VALUE)

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  private final long nanoTime;
+
+  /**
+   * Returns the {@code Time} with the given {@code timestamp}.
+   *
+   * @param timestamp the timestamp associated with this {@code Time}.
+   * @return the {@code Time} with the given {@code timestamp}.
+   */
+  static Time withTimestamp(Timestamp timestamp) {
+    return new Time(timestamp, 0);
```
Comment (sebright) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: Does zero mean that there is no `nanoTime`?  I think that there should be a comment, especially since there is a getter that returns the raw value.

File: core/src/test/java/com/google/instrumentation/common/TimestampTest.java [hunk 11]
```diff
@@ -36,20 +36,39 @@ public void testTimestampCreateInvalidInput() {
...
   }
 
   @Test
+  public void timestampAddNanos() {
+    Timestamp timestamp = Timestamp.create(1234, 223);
+    assertThat(timestamp.addNanos(0)).isEqualTo(timestamp);
+    assertThat(timestamp.addNanos(999999777)).isEqualTo(Timestamp.create(1235, 0));
+    assertThat(timestamp.addNanos(1300200500)).isEqualTo(Timestamp.create(1235, 300200723));
+    assertThat(timestamp.addNanos(1999999777)).isEqualTo(Timestamp.create(1236, 0));
+    assertThat(timestamp.addNanos(9876543789L)).isEqualTo(Timestamp.create(1243, 876544012));
```
Comment (a-veitch) [inline core/src/test/java/com/google/instrumentation/common/TimestampTest.java]: I think tests that use functions with int/long params should always include cases where max values are used. please add.

File: core_impl/src/test/java/com/google/instrumentation/trace/TimeConverterTest.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    assertThat(timeConverter.convert(time)).isEqualTo(timestamp);
+  }
+
+  @Test
+  public void convertTime_WithNanoTime() {
+    Timestamp timestamp = Timestamp.create(1234, 0);
+    TimeConverter timeConverter = new TimeConverter(timestamp, 1000);
+    assertThat(timeConverter.convert(Time.withNanoTime(1234))).isEqualTo(timestamp.addNanos(-234));
+    assertThat(timeConverter.convert(Time.withNanoTime(678))).isEqualTo(timestamp.addNanos(322));
+  }
```
Comment (sebright) [inline core_impl/src/test/java/com/google/instrumentation/trace/TimeConverterTest.java]: There should also be a test with non-zero nanoseconds.

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+package com.google.instrumentation.trace;
+
+import com.google.common.base.MoreObjects;
+import com.google.instrumentation.common.Timestamp;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Class that encapsulates one of values {@link Timestamp} or nanoTime. Used for internal
```
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: s/values//

File: core_impl/src/test/java/com/google/instrumentation/trace/TimeTest.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  public void withNanoTime() {
+    long nanoTime = 1234;
+    Time time = Time.withNanoTime(nanoTime);
+    assertThat(time.hasTimestamp()).isFalse();
+    assertThat(time.getTimestamp()).isNull();
+    assertThat(time.getNanoTime()).isEqualTo(nanoTime);
+  }
+
+  @Test
+  public void traceId_ToString() {
```
Comment (sebright) [inline core_impl/src/test/java/com/google/instrumentation/trace/TimeTest.java]: "traceId" looks like a typo.

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.google.instrumentation.common.Timestamp;
+import com.google.instrumentation.common.TimestampFactory;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * This class provides a mechanism for converting {@link Time} values to {@link Timestamp}. An
+ * instance of this class in initialized with a pair of timestamp/nanoTime measurements that wre
+ * taken at approximately the same time.
+ */
```
Comment (sebright) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: I think this Javadoc could also be expanded to make the class' purpose clearer.

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+  private final Timestamp timestamp;
+  private final long nanoTime;
+
+  /**
+   * Returns the {@code Time} with the given {@code timestamp}.
+   *
+   * @param timestamp the timestamp associated with this {@code Time}.
+   * @return the {@code Time} with the given {@code timestamp}.
+   */
+  static Time withTimestamp(Timestamp timestamp) {
```
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: suggest fromTimestamp would be a better name.

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java (hunk 1 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: fromNanoTime would be better name.

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+   *     false}.
+   */
+  boolean hasTimestamp() {
+    return timestamp != null;
+  }
+
+  /**
+   * Returns the timestamp associated with this {@code Time}.
+   *
+   * @return the timestamp associated with this {@code Time}.
```
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: Document this could return null.

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+   */
+  Timestamp getTimestamp() {
+    return timestamp;
+  }
+
+  /**
+   * Returns the nano time associated with this {@code Time}.
```
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: Document what return value of 0 indicates.

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java (hunk 2 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: s/in/is/
s/wre/are/

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.instrumentation.common.Timestamp;
+import com.google.instrumentation.common.TimestampFactory;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * This class provides a mechanism for converting {@link Time} values to {@link Timestamp}. An
+ * instance of this class in initialized with a pair of timestamp/nanoTime measurements that wre
+ * taken at approximately the same time.
```
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: This second sentence is implementation detail, not needed for usage of the class, suggest deleting it, and moving it to a // comment somewhere.

Also, you might want to replace "at approximately the" with "as close as possible to"

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+@Immutable
+final class TimeConverter {
+  private final Timestamp timestamp;
+  private final long nanoTime;
+
+  static TimeConverter now() {
+    return new TimeConverter(TimestampFactory.now(), System.nanoTime());
+  }
+
+  /**
```
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: Not a public function - needs javadoc?

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+ * This class provides a mechanism for converting {@link Time} values to {@link Timestamp}. An
+ * instance of this class in initialized with a pair of timestamp/nanoTime measurements that wre
+ * taken at approximately the same time.
+ */
+@Immutable
+final class TimeConverter {
+  private final Timestamp timestamp;
+  private final long nanoTime;
+
+  static TimeConverter now() {
```
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: No comment documenting this.

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java (hunk 2 above)
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: Done.

Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: Done.

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java (hunk 17 above)
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: Removed class.

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java (hunk 1 above)
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: Removed class.

Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: Removed class.

File: core_impl/src/main/java/com/google/instrumentation/trace/Time.java (hunk 3 above)
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/Time.java]: Removed class.

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java (hunk 20 above)
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: Don't know. We should decide on this.

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java (hunk 21 above)
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: We should decide what do we need for package-protected methods/classes.

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java (hunk 19 above)
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: Removed.

File: core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java (hunk 15 above)
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/trace/TimeConverter.java]: Hope is cleared.

File: core/src/main/java/com/google/instrumentation/common/Timestamp.java (hunk 9 above)
Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/common/Timestamp.java]: Done.

Comment (bogdandrutu) [inline core/src/main/java/com/google/instrumentation/common/Timestamp.java]: Done.

File: core/src/test/java/com/google/instrumentation/common/TimestampTest.java (hunk 11 above)
Comment (bogdandrutu) [inline core/src/test/java/com/google/instrumentation/common/TimestampTest.java]: Done.
